proof/refine/Move_R.thy

```diff
@@ -457,4 +457,20 @@ lemma bind_dummy_ret_val:
    od = do a; b od"
   by simp
 
+(* An "excluded middle" lemma for sc_at_ppred *)
+lemma sc_at_ppred_exm:
+  "sc_at g s \<Longrightarrow> sc_at_ppred p P g s = (\<not> sc_at_ppred p (\<lambda>x. \<not> P x) g s)"
+  by (clarsimp simp: sc_at_pred_n_def obj_at_def is_sc_obj pred_neg_def)
+
+lemma tcb_obj_at_equiv:
+  "tcb_at g s \<Longrightarrow> obj_at (P s) g s = (\<forall>ko. ko_at ko g s \<and> is_tcb ko \<longrightarrow> P s ko)"
+  by (clarsimp simp: obj_at_def is_tcb)
+
+(* FIXME RT: replace get_ep_valid_ep with this better version *)
+lemma get_ep_valid_ep'[wp]:
+  "\<lbrace> valid_objs and ep_at ep \<rbrace>
+   get_endpoint ep
+   \<lbrace> valid_ep \<rbrace>"
+  by (wpsimp simp: ep_at_def2 valid_ep_def2 simp_del: valid_simple_obj_def wp_del: get_ep_valid_ep)
```

Comment:
Do you mind if I leave this as a cleanup task? I've been trying to keep my PRs restricted to refine because working across two images tends to slow down the process (I have to rebuild things all the time). But I will do the replacement if you are very sure you want it now.

Reply:
Ah, but I see there is other overlap with AInvs on other comments. Hmm.